examples/ladder_nets/ladder_nets.py

```diff
@@ -0,0 +1,436 @@
+                                      name='{}_{}'.format(name, dense_lname))
+
+        # 1. batchnormalize without learning -> goes to combinator layer
+        l_name = '{}_{}'.format(name, nbatchn_lname)
+        net[nbatchn_lname] = BatchNormLayer(net.values()[-1], alpha=0.1, name=l_name,
+                                            beta=None, gamma=None)
+
+        if shared_net is None:
+            # for dirty encoder -> add noise
+            net[noise_lname] = GaussianNoiseLayer(net.values()[-1],
+                                                  sigma=p_drop_hidden,
+                                                  name='{}_{}_'.format(name,
+                                                                       noise_lname))
+
+        # 2. scaling & offsetting batchnormalization + noise
+        l_name = '{}_{}'.format(name, lbatchn_lname)
+        # offset by beta
+        net[lbatchn_lname + '_beta'] = BiasLayer(net.values()[-1], b=beta,
+                                                 name=l_name + '_beta')
+
+        if gamma is not None:
+            # if not rectify, scale by gamma
+            net[lbatchn_lname + '_gamma'] = ScaleLayer(net.values()[-1],
+                                                       scales=gamma,
+                                                       name=l_name + '_gamma')
+        # apply activation
+        if i < len(num_hidden) - 1:
+            act_name = 'enc_activation_{}'.format(i)
+            net[act_name] = NonlinearityLayer(net.values()[-1],
+                                              nonlinearity=activation,
+                                              name='{}_{}'.format(name, act_name))
+
+    # classfication layer activation -> softmax
+    net['enc_softmax'] = NonlinearityLayer(net.values()[-1], nonlinearity=softmax,
+                                           name='{}_enc_softmax'.format(name))
+
+    return net['enc_softmax'], net
+
+
+def build_decoder(dirty_net, clean_net, num_nodes, sigma, combinator_type):
+    L = len(num_nodes) - 1
+
+    # dirty_enc_dense_1 ... z_L
+    z_L = dirty_net['enc_noise_{}'.format(L)]
+
+    # batchnormalized softmax output .. u_0 without learning bn beta, gamma
+    dirty_net['u_0'] = BatchNormLayer(dirty_net.values()[-1], beta=None,
+                                      gamma=None, name='dec_batchn_softmax')
+
+    # denoised latent \hat{z}_L = g(\tilde{z}_L, u_L)
+    comb_name = 'dec_combinator_0'
+    dirty_net[comb_name] = CombinatorLayer(z_L, dirty_net['u_0'],
+                                           combinator_type=combinator_type,
+                                           name=comb_name)
+
+    # batchnormalize denoised latent using clean encoder's bn mean/inv_std without learning
+    enc_bname = 'enc_batchn_{}_norm'.format(L)
+    bname = 'dec_batchn_0'
+
+    to_stats_l = clean_net[enc_bname]
+    to_norm_l = dirty_net[comb_name]
+    dirty_net[bname] = SharedNormLayer(to_stats_l, to_norm_l)
```

Comment:
Now this removes the mean and divides by the standard deviation that was also used in the encoding step -- does this make sense? Shouldn't the decoder be doing the reverse? You also have standard BatchNormLayers in the decoder, maybe you don't need the SharedNormLayers at all? (Disclaimer: I haven't looked at your code or the paper in detail, I'm just wondering. I'm happy to learn why it is implemented the way it is.)